packages/core/src/browser/common-frontend-contribution.ts

```diff
@@ -350,6 +364,10 @@ export class CommonFrontendContribution implements FrontendApplicationContributi
         this.contextKeyService.createKey<boolean>('isWindows', OS.type() === OS.Type.Windows);
         this.contextKeyService.createKey<boolean>('isWeb', !this.isElectron());
 
+        this.pinnedKey = this.contextKeyService.createKey<boolean>('activeEditorIsPinned', false);
```

Comment:
Shouldn't the functionality be specific to `editors` only, what does it mean to have a pinned `widget` (ex: scm).
If it's the case, the functionality should not be added to `@theia/core`.

Reply:
1. Common frontend contribution seems to be the place where all editor-tabs commands reside. Well, pinning widgets might help those who accidentally middle-click them although that doesn't quite make sense typically.
2. Nice catch, should be decorated I guess, but need to choose how should it look.
3. That's close button of tab, which is managed by Phosphor (not sure if I'm right), need to somehow intercept event and do unpin instead of close.
4. It does... I didn't really think about sidebar, arrangement code is specific to editor tabs probably

Reply:
I think we can mitigate those problems (1, 2, 4) by removing pin/unpin commands for non-editor widgets. 3 needs investigation on opportunities

Reply:
@sh7dm initially I think we can implement pinning of editors only (`@theia/editor` extension), and likely only in the main-area for simplicity.

Reply:
> Shouldn't the functionality be specific to `editors` only, what does it mean to have a pinned `widget` (ex: scm).
> If it's the case, the functionality should not be added to `@theia/core`.

This is always a tricky terminological gap to bridge between Theia and VSCode, since in VSCode, 'editor' means 'editor or webview,' (see screenshot for a pinned webview) but also implies 'any main-area widget'. Since Theia is much more flexible with what goes in what area, allowing editors and webviews to appear outside the main area and other widgets to appear in it, I'm not sure restricting the pinning functionality to editors makes much sense. I would be inclined to extend the ability to any closeable widget on any tabbar that allows 'close all' or 'close other' functionality (all of them, I think). Otherwise the logic to achieve strictly VSCode-like behavior when so much else is different can get torturous.

![image](https://user-images.githubusercontent.com/62660806/123108985-80261500-d400-11eb-934c-8c223f59e49a.png)

Reply:
So we can fix pinned widgets placement on non-main views and decorate those, I think that would be really better and help avoiding edge cases @colin-grant-work described.

`activeEditorPinned` name was taken from VSCode keyboard shortcuts and should really be `activeWidgetPinned` since it works for everything.

Reply:
Let's go ahead with that :+1: Let's make it available for any closeable widget, and keep it generic (ex: `activeWidgetPinned`) like you've proposed would be a fine `contextKey`. The only issue would be determining how to properly decorate views not in the main-area (likely only sidebar has issues at the moment), and solving the issue of views re-arranging oddly in the sidebars.

Reply:
Re: the context key name, we should probably either keep `activeEditorPinned` or implement both `activeEditorPinned` and `activeWidgetPinned` since plugin `when` clauses will expect `activeEditorPinned`.

Reply:
My issue was having editor references in `@theia/core` but you are right about plugins using the same `contextKey` so it should be implemented.